### DISCUSSION TITLE
CUDA implementation for PhysX

### DISCUSSION THREAD

**Discussion by liandris** [vosen/ZLUDA#31] [Ideas]
This might seem not fitting, however I think here is the best place since ZLUDA is the only project I've seen so far that comes close to how I think it should work. More specifically - re-implemented nvcuda.dll file that can be placed where needed.

### What is this about?

PhysX is a "realtime physics engine middleware" as its [wikipedia page](https://en.wikipedia.org/wiki/PhysX) says. Developed by NovodeX AG which was acquired by Ageia in 2004, and Ageia was acquired by Nvidia in 2008. Both Ageia and Nvidia used it to promote their hardware. Nvidia had to rewrite parts of it so it would use CUDA on their graphics cards, so for some years it was a subject of criticism and controversy in some communities. PhysX can fully work on CPU and only use GPU mode when supported Nvidia card is available. Basically, Nvidia did all they could to promote PhysX among game developers but did nothing to help them properly optimize it, which made their hardware look more appealing due to big performance difference.

That changed a bit with introduction of PhysX v3 in 2011 which was a significant rewrite, and it did bring performance improvements to CPU mode. Right now it's not easy to find information about whether or not some new game uses PhysX v3 (or above) and supports GPU mode for it - I wouldn't be surprised if there is only a couple of such games. This may also be thanks to collaboration with Epic Games - PhysX was well integrated with Unreal Engine since UE3, but UE4 does not offer hardware acceleration support for its PhysX v3/v4 features. Even though its GPU mode is still there, it was rarely or never used, because most features run well on CPU.

Since its v4, PhysX became open source and it's available here: https://github.com/NVIDIAGameWorks/PhysX

The problem still remains with those games that were released with PhysX v2 included. There is no way to move them to newer PhysX versions to improve performance, and all those added effects still massively impact performance even on modern CPUs when enabled. There is also a huge contrast in visuals of most of such games when compared between PhysX on/off modes - since it mostly served marketing purposes, it was rarely implemented properly. In some cases it's a choice between excessive amounts of smoke/particles/debris and zero amounts of those. In other cases the game might allow to switch between 2-3 presets or manipulate some settings through configuration files, but in absolute most cases it will not be enough to make the game playable while PhysX is enabled.

A rough list of affected games can be found [here](https://web.archive.org/web/20200706132532/http://physxinfo.com/). There is another list of [wikipedia](https://en.wikipedia.org/wiki/List_of_games_with_hardware-accelerated_PhysX_support) but it's much less accurate and sometimes misleading.

### Why care about video games?

It could be a significant milestone for a project like this. It gives one more reason to make it work on AMD (or any other) GPU. And it certainly will be appreciated by PC community.

### Can it even work?

It's unknown. But at least PhysX v2 calculations that are done through CUDA do not depend on anything that happens in game - CUDA device does not need to have access to 3D renderer.

There are multiple problems that can be outlined though.
- Most affected games are 32-bit, and ZLUDA is 64-bit only (because oneAPI Level Zero is 64-bit). Either it has to be changed to use something else, or it will have to involve interprocess communication (IPC).
- oneAPI Level Zero is Intel-only. Something else must be used in order to support wide range of devices. Possibilities include OpenCL and DirectCompute. Also there is [VUDA project](https://github.com/jgbit/vuda) that is based on Vulkan.
- There are several iterations of tech that include PhysX, including APEX and GameWorks. There is also a number of dll modules that interact between the game and CUDA, like nvapi and cudart. Such modules may include more stuff that would prevent a 3rd party CUDA implementations from working. Their functionality is not documented, however it should be possible to solve such issues case by case. However from what I understand it's only true for a few games.

Theoretically it's already possible to test ZLUDA as a CUDA processor for PhysX on a real game - I think the only fitting 64-bit game is Batman Arkham Knight. However it uses PhysX v3 which may require access to 3D renderer for certain bits (just my assumption), so it might not be the frendliest case. I can't test this myself for now since I don't own a supported Intel GPU.

@vosen suggested posting the details here. I guess this can be considered a milestone draft or a discussion place for anyone interested.

**Comment by vosen** [comment]
Continuing discussion from #72 

I had a look into a 32 bit game using PhysX. Specifically at Sacred 2. I will not be able to do it myself, but if there's someone out there who wants to take a stab, here's my thoughts:

- ROCm also is 64 bit only. The most realistic implementation path is to create a 32 bit "proxy" CUDA implementation.
  - Proxy implementation would not run any GPU code by itself, rather it would send all the function calls to a 64 bit server which would only accept commands from the proxy and run GPU computation
  - 64 bit handles from server can be eqasily converted to 32 bit handles with lookup tables
  - Data is the problem. I've extracted PhysX PTX and while the instruction mix does not seem to be too crazy (loads of ld, st, floating point math, normal GPU stuff), pointers inside kernels are a problem. I think a viable solution would be to do a fake allocator: preallocate a one big memory buffer (let's say 1 GB, but might start smaller grow dynamically) and convert all the pointers as offsets into the buffer. That would require introducing special "32 bit on 64 bit mode" for global and generic memory access
  - PhysX's Dark API use is IIRC mapped out, there might be one or two functions that are stubbed out but have a well-understood mechanism

- An alternative would be to sort of rebuild ZLUDA on top of 32 bit Vulkan. Maybe Vulkan with extensions can support enough CUDA features to run PhysX? But it's a ton of work since you will have to start from scratch with host code and convert GPU code emission to emit SPIR-V instead of LLVM

**Reply by mirh** [reply]
@Mystro256 [looked](https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/MPOPPUJEEG44ODBW6NRAOCVH6ARRXFV5/) into it in the past.. not sure if fixing x86 rocm-runtime is just into "too useless, couldn't be bothered" or "a full time engineer really has to waste their ass" territory. 

Anyhow, putting aside that wine's WoW64 thunking mechanism should totally eventually support this 32-to-64 scenario (even though idk how effortful it would be to do that for all the [nvapi](https://github.com/wine-staging/wine-staging/tree/v9.2/patches/nvapi-Stub_DLL) and [cuda](https://github.com/wine-staging/wine-staging/tree/v9.2/patches/nvcuda-CUDA_Support) apis required).. You could just try any of the 64-bit native games with [hardware effects](https://www.youtube.com/playlist?list=PL4FII4B-zM0f9w-s3jnjpp46DiU6zDoCA)? Like.. I don't know Batman: Arkham Knight or the metro reduxs.

**Comment by Deus-nsf** [comment]
I'm still hoping one day that Mirror's Edge 2008 could support it, most of the games supporting PhysX can actually brute force the computation through the CPU, but for some reason Mirror's Edge 2008 uses the CPU in single thread when processing PhysX effects like glass particles breaking etc... Making it absolutely unusable with anything else than an Nvidia card (we're talking +- 15 FPS with top of the line CPUs).

**Reply by liandris** [reply]
Are you sure you evaluated other games properly? From my understanding there is little to none games that actually use multi-core CPU implementation of PhysX, so I expect them to be single-threaded. I remember trying some of them and none come close to being playable, just like Mirror's Edge. A huge factor is actually playing through the game, for example in Mirror's Edge you may actually get away with first few levels thinking framerates are plausible but at later levels you'll find effects that rely on PhysX in much larger quantities. That would be single-digit FPS level of performance.

**Comment by Koth-87** [comment]
I absolutely endorse this idea, and while I don't have any real dev skills, I have a 7900 XTX and Arkham Knight (and other hardware-PhysX games) so I can test stuff and provide data.

**Comment by abasgames** [comment]
might be the time

**Reply by Hyenadae** [reply]
So something goofy like 32bit VLUDA could theoretically work if we had enough... well, crazy people

**Comment by Iordan-Iulian-Bogdan** [comment]
Seems feasible for games using older versions, I could look into this but there's another issue besides the 32bit compatibility, as far as I can see from the titles I checked they don't explicitly use the CUDA DLLs in their files, I am assuming they use system level DLLs which come with Nvidia drivers.

Can anyone with an Nvidia card test to see if deleting nvcuda.dll and nvml.dll causes games to not work ? I don't have access to one at the moment, I think they're supposed to be in System32 and \NVIDIA Corporation\NVSMI.

**Reply by vosen** [reply]
If you want to tackle this the open a dedicated Github issue or open a discussion on #devtalk on ZLUDA Discord.
IIRC 32bit physx games do load nvcuda.dll, not sure about nvml.dll. The mechanism for loading nvcuda.dll or nvml.dll should not matter: zluda_with.exe should handle it (and if it does not the first step is to fix it)

**Comment by Iordan-Iulian-Bogdan** [comment]
I thought about this some more and I am wondering if instead of trying to rebuild something like ZLUDA for 32bit compatibility perhaps a simpler stop gap solution for older games might be to actually just use a CPU backend instead. 

I suspect these CUDA kernels would run fast enough to offer a good boost to performance once recompiled/interpreted to target a multi threaded x86-64 backend on modern CPUs, I tried looking around if anyone has already made something like this but most projects are very old/dead or I can't find performance metrics to see if it would be worth looking into. If anyone knows of one please tell.

**Reply by liandris** [reply]
Some articles suggest those old implementations use x87, which is likely very difficult or impossible to recompile into proper optimized x86. Making the code work in a multi-threaded manner is a whole another challenge.

I too thought that at some point CPUs will evolve to the point of making stuff like Mirror's Edge playable. Alas, 15 years have passed and it didn't improve at all. It seems the CPU part was intentionally coded to avoid good performance.

Projects trying to emulate PhysX alone will likely fail to meet expectations. Projects that aim to emulate CUDA, though, should have more support and be more successful, and it would eliminate the need to touch the old clunky PhysX code.

**Reply by Iordan-Iulian-Bogdan** [reply]
Some very early versions of CUDA actually allowed you to compile to x86 but I don't recall if they were multi threaded/vectorized or if the performance would be any good on today's CPUs. GPUOcelot looks like it might be what I am looking for, I'll look into it.

**Reply by liandris** [reply]
How CUDA allows you to compile is probably irrelevant to PhysX x87 CPU code. Its old versions were never open sourced.

Multi-thread feature for CPU path appeared quite late when it comes to games using PhysX GPU path, it was not enabled by default (it was said developers needed to do some extra work for it to work) and I don't know if any game actually used it. FluidMark shows it can improve performance, but it had to be updated in order to be able to use it at all. Also my impression is that even if you can enable it for some games, it may bring (very roughly) 200% improvement or so in heavy scenarios. 200% of 10 fps is only 20 fps.

Speaking of which, FluidMark is probably a good test case candidate for a possible 32-bit CUDA implementation.

**Reply by vosen** [reply]
ZLUDA CPU backend is possible, but that would be a much bigger project and would give you worse performance than 32bit CUDA bridge to 64bit CUDA

**Comment by chkuendig** [comment]
With the relaunched ZLUDA focusing on Linux there might be a somewhat easier path to this:

Wine already has support for running [PhysX accelerated on top of the linux cuda libraries](https://gitlab.winehq.org/wine/wine-staging/-/wikis/Configuration/PhysX). Combined with the ["new WoW64" which can run 32bit apps in 64bit unix processes]( https://gitlab.winehq.org/wine/wine/-/releases/wine-9.0#wow64), wouldn't that mean that making sure the 64bit linux build of ZLUDA is found by wine is enough? Presumably there will be bugs as 3 unstable components are combined, but naively this seems like the easiest way?

PS:
-  ~[PTX comes in 32bit and 64bit variants](https://stackoverflow.com/questions/20624635/cuda-ptx-32-bit-vs-64-bit), that could be an additional issue if old PhysX uses the 32bit ISA and the ZLUDA ptx compiler doesnt support it.~ _[not an issue](https://discord.com/channels/1273316903783497778/1273316903783497781/1349395117446008853)_
- For it to be usable without endless tinkering, proton would first have to support "new WoW64" (https://github.com/ValveSoftware/Proton/issues/6889)

**Reply by liandris** [reply]
>Wine already has support for running [PhysX accelerated on top of the linux cuda libraries](https://gitlab.winehq.org/wine/wine-staging/-/wikis/Configuration/PhysX)

It seems that support does not involve emulating anything - you still need Nvidia GPU for it to work as intended. This is not better than having a Windows machine with Nvidia GPU.

>Combined with the ["new WoW64" which can run 32bit apps in 64bit unix processes](https://gitlab.winehq.org/wine/wine/-/releases/wine-9.0#wow64)

If it's anything similar to the original WoW64 it will not act as a layer between 32-bit games and 64-bit CUDA-like runtimes.

**Reply by mirh** [reply]
It is not similar, that's the thing they added for osx in order to run 32-bit exes despite the 64-bit only OS requirement.
But ZLUDA shouldn't really have to think about that added complexity.

**Reply by chkuendig** [reply]
> It seems that support does not involve emulating anything - you still need Nvidia GPU for it to work as intended. 

There's no need to emulate anything, the whole goal of ZLUDA is to be a drop-in replacement for the Cuda libraries (which is why they even have the same filenames). 

> It is not similar, that's the thing they added for osx in order to run 32-bit exes despite the 64-bit only OS requirement. But ZLUDA shouldn't really have to think about that added complexity.

If I understood correctly, ZLUDA can't be built for 32bit (and probably shouldn't anyway - it's 2025), so having the game and Physx run in a 64bit process would help a lot (since that can then use 64bit libraries on the linux side of things).

**Reply by mirh** [reply]
OP was about the old intel-focused work (and honestly I'm somewhat skeptical that was even correct, because *at most* it should be just a matter of having to build yourself some libraries).
It doesn't necessarily apply to ROCm (even though that has some proprietary components here and there IIRC and it might be more of the case).

**Reply by liandris** [reply]
> a drop-in replacement for the Cuda libraries

That's what I would call an emulation. It involves using hardware not originally intended for a certain feature, which is a non-Nvidia GPU in this case. The mentioned feature of Linux libraries does not allow that, so I assume it's useless for this case.

> If I understood correctly, ZLUDA can't be built for 32bit (and probably shouldn't anyway - it's 2025), so having the game and Physx run in a 64bit process would help a lot (since that can then use 64bit libraries on the linux side of things).

Having a 32-bit game in a 64-bit process (whatever that means) will not make it use 64-bit CUDA runtimes because they are not a part of standard Windows installation.

> OP was about the old intel-focused work

Only because ZLUDA was the only drop-in replacement for CUDA. My interest is more in the emulating CUDA libraries and their functions, less in code specific to Intel hardware. With some time and effort it could start working in vendor-agnostic environments like OpenCL, DirectCompute etc.

**Comment by mirh** [comment]
> Having a 32-bit game in a 64-bit process (whatever that means) will not make it use 64-bit CUDA runtimes 

Yes that's the whole point.
*If any* you are right on the Windows point, in the sense that this is only a linux thing.

> My interest is more in the emulating CUDA libraries and their functions, less in code specific to Intel hardware. 

SYCL isn't intel specific, but anyway what you are talking about isn't "emulation" (you know? wine is not an emulator). It's simply a compatibility layer or a reimplementation.

**Reply by liandris** [reply]
> > Having a 32-bit game in a 64-bit process (whatever that means) will not make it use 64-bit CUDA runtimes
> 
> Yes that's the whole point. _If any_ you are right on the Windows point, in the sense that this is only a linux thing.

This shouldn't differ between Windows and Linux. I don't expect the resulting successful project to somehow allow proper non-Nvidia PhysX support exclusively on Linux.

Wine's WOW64 mode only covers default Windows libraries as per description. CUDA libraries are not touched.

> but anyway what you are talking about isn't "emulation" (you know? wine is not an emulator). It's simply a compatibility layer or a reimplementation.

These semantics shouldn't matter. Either some part of a project involves touching or re-implementing CUDA libraries or doesn't. Wine doesn't and therefore it won't help. There is no way around it from what I see.

**Reply by mirh** [reply]
> Wine's WOW64 mode only covers default Windows libraries as per description. CUDA libraries are not touched.

And FWIW vanilla wine doesn't even cover any nvidia library to begin with.
But it's not rocket science. It's just about wrapping and padding every call between bitness, and I suppose it shouldn't take too much to extend the normal nvcuda patches.

> These semantics shouldn't matter. 

They totally matter.
Though I guess it just depends on whether you imagine cuda to be an api or an instruction set.

**Comment by chkuendig** [comment]
For what it's worth, I tried getting this to work with Proton-GE on my steamdeck. What I did:
- Fixed the [cuda patches in wine-staging](https://github.com/wine-staging/wine-staging/tree/cab93f47b8d095eb968bb3c7debf9117a91307ce/patches/nvcuda-CUDA_Support) so they can be used with proton 9.
- Removed the restriction in dxvk-nvapi limiting cuda to the proprietary driver: https://github.com/jp7677/dxvk-nvapi/blob/e870088aa074b031b0f13d7d37500ca9d4834768/src/nvapi_gpu.cpp#L657 (FYI, the discussion at https://github.com/jp7677/dxvk-nvapi/issues/27 is pretty enlightening)

- ROCm libraries copied from steamos/holo repo to proton folder (/home/deck/.local/share/Steam/compatibilitytools.d/proton-cuda/files/lib64/) so they can be accessed from within the pressure vessel container
  - from comgr: `sudo cp /opt/rocm/lib/libamd_comgr.so* /home/deck/.local/share/Steam/compatibilitytools.d/proton-cuda/files/lib64/`
  - from rocminfo: `sudo cp /opt/rocm/lib/libhsa-runtime64.so* /home/deck/.local/share/Steam/compatibilitytools.d/proton-cuda/files/lib64/`
  - from numactl: `sudo cp /usr/lib/libnuma.so* /home/deck/.local/share/Steam/compatibilitytools.d/proton-cuda/files/lib64/`
  - from hsakmt-roct: `sudo cp /opt/rocm/lib/libhsakmt.so*   /home/deck/.local/share/Steam/compatibilitytools.d/proton-cuda/files/lib64/`
  - using a custom-built HIP runtime to work around this bug: [rocm-opencl-runtime#2](https://gitlab.archlinux.org/archlinux/packaging/packages/rocm-opencl-runtime/-/issues/2 ) - using the same version as in the holo repos (6.0.3). Built as outlined at https://rocm.docs.amd.com/projects/HIP/en/latest/install/build.html 
- Tested with ZLUDA 4 and 3 (won't matter much as you'll see below), copied to the same folder
- Experimented with various flags, mostly using `DXVK_NVAPI_LOG_LEVEL=trace DXVK_NVAPI_ALLOW_OTHER_DRIVERS=1 DXVK_CONFIG=dxgi.hideAmdGpu = False HSA_OVERRIDE_GFX_VERSION=10.3.0 PROTON_LOG=1 %command% ` 
  - (HSA_OVERRIDE_GFX_VERSION in anticipation of the [Van Gogh GPU not being fully supported](https://element84.com/machine-learning/running-rocm-5-4-2-onnx-and-pytorch-on-a-steamdeck/))

**Results:**
Limited testing to 64bit apps to keep it simple and avoid any WoW64 pitfalls (solving that would be straightforward).

- Metro Exodus: Allows me to enabled "Advanced Physx" and Hairworks, but never loads nvcuda.dll. Crashes sporadically.
- Arkham Knight: This loads nvcuda.dll and also fetches devices from nvapi. Unfortunately hangs somewhere in ibvulkan_radeon.so (the userspace libraries of RADV) with a loop throwing exceptions. I don't have the knowhow or energy to symbologicate the crash. I assume the reason it's even loading nvcuda.dll is unrelated to PhysX but the fact that it uses some Gameworks effects that directly use CUDA (and can't be activated on non-Nvidia cards - there's no CPU fallback). This was the app I patched dx-nvapi for so the device list would be populated (it failed at that intially)
- Entropy Centre: Warns about an unsupported driver (DXVK_NVAPI_DRIVER_VERSION doesn't have an effect, so it seems it's getting the version from Vulkan/D3D not NVAPI), doesn't have a physx option and never loads nvcuda.dll.
- A Way Out: Doesn't launch, hangs somewhere very early (proton log is empty)

**Conclusion:**
I think getting this to work (PhysX with ZLUDA) will need extensive changes to DXVK and dxvk-nvapi to fully impersonate the Nvidia Drivers. It might be worthwhile to try out the NVAPI patches in Wine-staging (since they stub everything without relying on the device drivers and apparently get PhysX working), but it's not clear they interoperate well with DXVK. 

Also, according to Nvidia, [driver signatures of nvcuda.dll are checked](https://github.com/NVIDIAGameWorks/PhysX-3.4/blob/5e42a5f112351a223c19c17bb331e6c55037b8eb/PhysX_3.4/release_notes.html#L1447):
> The PhysXCommon/64.dll, nvcuda.dll and PhysXUpdateLoader/64.dll are loaded and checked for the NVIDIA Corporation digital signature. The signature is expected on all NVIDIA Corporation provided dlls. The application will exit if the signature check fails.
                
So that would probably also have to be worked around eventually.

PS: Since there's been a bit of a misunderstanding above about why I proposed doing this with proton: Trying to use ZLUDA to replace the CUDA libs on WIndows will be even more complicated. You'll have to do all of the above but achieve it by patching binary libraries (DirectX) and rewriting them or porting over their proton-equivalent (NVAPI).

**Reply by liandris** [reply]
> Limited testing to 64bit apps

- All listed games use PhysX v3.x.
- It seems only Metro Exodus and Arkham Knight have GPU PhysX effects. The other two only rely on CPU mode.

All previous 32-bit games appear to use PhysX v2. Compared to v3, its architecture is different as it relies on separate "PhysX System Software". And I expect v2 environment to be more prone to modifications. Nvidia's little war with 3rd-party patch (which allowed using secondary Nvidia GPU together with primary non-Nvidia GPU) where they introduced more integrity checks started and ended before PhysX v3 was shipped with games. Many PhysX v3 games ship with "APEX" libraries that may have additional checks. Arkham Knight involved a handful of other Nvidia-exclusive features so it may be more difficult to handle.

>You'll have to do all of the above but achieve it by patching binary libraries (DirectX) and rewriting them or porting over their proton-equivalent (NVAPI).

Likely not DirectX, but PhysX and CUDA. I may be wrong but NVAPI seems to be much less relevant to how PhysX v2 games decide between CPU and GPU modes.

**Reply by mirh** [reply]
Even v3 (sometimes) relies on PSS.
Also, they removed most hybrid physx checks in 2016 (and I'm skeptical these were implemented into the single PhysXCore dlls).

**Reply by Aytackydln** [reply]
NVIDIA released the source code for GPU PhysX:

https://github.com/NVIDIA-Omniverse/PhysX/releases/tag/107.0-physx-5.6.0

How useful is it?

**Comment by vosen** [comment]
Closing because it's becoming a reality as seen in #651 (and discussed in https://vosen.github.io/ZLUDA/blog/zluda-update-q1q2-2026/#physx-pre-alpha)
